### PR TITLE
Fixed #24835 - exists() incorrect after annotation with Count()

### PR DESCRIPTION
**Pull request by pwmarcz**
https://code.djangoproject.com/ticket/24835

I followed the same approach as in `QuerySet._values()`. Seems hacky but I couldn't figure out a better way.

### CODE REVIEW

**Comment by charettes** [comment]
Buildbot test this please.

**Comment by jarshwah** [comment]
I think the approach you've taken is mostly correct. It's unfortunate that similar processing has to be done in so many places, but that's a task for another time.

I feel like the majority of the work should be performed by `db.sql.query.has_results` rather than `db.query.exists`. Is there a technical reason for this?

**Comment by akaariai** [comment]
I second jarshwah's opinion - the logic should be in has_results(), not in exists(). (Seems like the ._values() logic should also be in sql.Query class, but this doesn't need to be solved here). We could even implement a dedicated helper method to be used both from ._values() and has_results().

**Comment by timgraham** [comment]
Please also add 1.8.3 release notes.

**Comment by pwmarcz** [comment]
>I second jarshwah's opinion - the logic should be in has_results(), not in exists().
I moved it to has_results().

>Please also add 1.8.3 release notes.
Done.

File: docs/releases/1.8.3.txt [hunk 1]
```diff
@@ -9,4 +9,7 @@ Django 1.8.3 fixes several bugs in 1.8.2.
 Bugfixes
 ========
 
+* Fixed `exists()` returning incorrect results after annotation with `Count()`
```
Comment (timgraham) [inline docs/releases/1.8.3.txt]: Need double backticks to have an effect on the rendered docs.

**Comment by timgraham** [comment]
Could you rebase and squash your commits? Thanks.

**Comment by pwmarcz** [comment]
Rebased on master and squashed.

As a rule, should I do that with every update or only at the end of review process?

**Comment by timgraham** [comment]
For small patches, squashing after each update is fine. For larger patches, it can be helpful to review only the changes. 

buildbot, add to whitelist.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by jarshwah** [comment]
:shipit: LGTM

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -489,6 +489,9 @@ def has_filters(self):
     def has_results(self, using):
         q = self.clone()
         if not q.distinct:
+            if q.group_by is True:
```
Comment (timgraham) [inline django/db/models/sql/query.py]: do we need `is True` here?

**Comment by jarshwah** [comment]
I've opened a new ticket for cleaning up the _values() method and potential factoring with has_results here: https://code.djangoproject.com/ticket/24854

File: django/db/models/sql/query.py (hunk 2 above)
Comment (jarshwah) [inline django/db/models/sql/query.py]: Yes, because it can also be "truthy" by containing a list of fields. True means all fields.

**Comment by timgraham** [comment]
merged in 801a84ae329a24d9adf01b700429fe8f1285b2b8, thanks!

**Pull request closed without merging**